### PR TITLE
[pdist] Fix process crash in pdist backward for zero-row inputs

### LINKED ISSUE #197099: Backward of `F.pdist` on an input with zero rows crashes the process (SIGSEGV on Linux, integer division by zero on Windows), no Python exception
### 🐛 Describe the bug

**Summary.** The backward of `F.pdist` on an input with zero rows does not fail gracefully: the forward returns an empty `(0,)` tensor, and `.backward()` terminates the interpreter — SIGSEGV on Linux, `STATUS_INTEGER_DIVIDE_BY_ZERO` on Windows — with no Python exception (torch 2.10.0 through 2.14.0 and nightly); inputs with one or two rows are fine and CUDA is unaffected. A user whose batch happens to be empty loses the process rather than getting an error. Because the CPU `pdist_backward` kernel divides by the number of rows without guarding zero, the crash is decided by the input shape alone, and torch.compile reaches the same kernel.

`torch.nn.functional.pdist` accepts an input with zero rows and returns an empty `(0,)` tensor, but calling `.backward()` on that result terminates the interpreter: on Linux the process is killed by **SIGSEGV**, on Windows it exits with `0xC0000094` (`STATUS_INTEGER_DIVIDE_BY_ZERO`). No Python exception is raised, so a training loop whose batch happens to be empty dies without a traceback. Inputs with one row (zero pairs) and two rows are fine, so the backward kernel does not guard `n == 0`. Reproduced on torch 2.14.0+cpu (Windows) and torch 2.10.0+cpu (Linux, Kaggle).

Minimal reproducer (run each case in a fresh interpreter; the failing ones kill the process):

```python
import torch

n, d = 0, 4
x = torch.randn(n, d, requires_grad=True)
y = torch.nn.functional.pdist(x)
print("forward ok, output shape", tuple(y.shape))     # (0,)
y.sum().backward()                                    # <-- process dies here for n == 0
print("backward ok, grad shape", tuple(x.grad.shape))
```

Driver that runs the four cases in subprocesses:

```python
import subprocess, sys
CASE = '''
import sys, torch
n, d = int(sys.argv[1]), int(sys.argv[2])
x = torch.randn(n, d, requires_grad=True)
y = torch.nn.functional.pdist(x)
print("forward ok, output shape", tuple(y.shape), flush=True)
y.sum().backward()
print("backward ok, grad shape", tuple(x.grad.shape), flush=True)
'''
for n, d in ((0, 1), (0, 4), (1, 4), (2, 4)):
    r = subprocess.run([sys.executable, "-c", CASE, str(n), str(d)], capture_output=True, text=True)
    print(f"pdist on ({n}, {d}):", " | ".join(r.stdout.split("\n")[:2]),
          "" if r.returncode == 0 else f"<-- process died, exit code {r.returncode & 0xFFFFFFFF:#x}")
```

Output, Linux (Kaggle CPU notebook, torch 2.10.0+cpu, Python 3.12.13, Linux 6.12.90 / glibc 2.35):

```
torch 2.10.0+cpu | Linux-6.12.90+-x86_64-with-glibc2.35 | 3.12.13
pdist on (0, 1): forward ok, output shape (0,)   <-- process killed by signal 11 (SIGSEGV)
pdist on (0, 4): forward ok, output shape (0,)   <-- process killed by signal 11 (SIGSEGV)
pdist on (1, 4): forward ok, output shape (0,) | backward ok, grad shape (1, 4)
pdist on (2, 4): forward ok, output shape (1,) | backward ok, grad shape (2, 4)
```

Output, Windows 11 (torch 2.14.0+cpu, Python 3.14.7):

```
torch 2.14.0+cpu | Windows-11-10.0.26200-SP0 | 3.14.7
pdist on (0, 1): forward ok, output shape (0,)   <-- process died, exit code 0xc0000094
pdist on (0, 4): forward ok, output shape (0,)   <-- process died, exit code 0xc0000094
pdist on (1, 4): forward ok, output shape (0,) | backward ok, grad shape (1, 4)
pdist on (2, 4): forward ok, output shape (1,) | backward ok, grad shape (2, 4)
```

On CUDA (Tesla T4, same torch 2.10.0+cu128) the backward returns a `(0, 4)` gradient normally, so the crash is in the CPU `pdist_backward` kernel only.

Still present on nightly torch 2.15.0.dev20260911+cu130 (nightly, git fb8ff0f1, Kaggle: Tesla T4 + Linux CPU), Linux CPU: the child process dies with exit code -11 (SIGSEGV) right after `forward ok (0,)`; CUDA backward still returns a `(0, 4)` gradient.

Expected: either the backward returns a zero gradient of shape `(0, d)` (consistent with the forward accepting the input), or the forward rejects zero-row inputs with a `RuntimeError`. A hard crash is never acceptable.

Observed: the interpreter is killed inside `pdist_backward` (CPU) — SIGSEGV on Linux, integer division by zero on Windows, i.e. the `n == 0` case is not handled. `torch.compile` does not help since the compiled path calls the same kernel.

Found by a differential harness that runs every OpInfo sample forward + backward on edge-shape variants (here: the batch dimension resized to 0).

### Versions

```
PyTorch version: 2.14.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro for Workstations (10.0.26200 64-bit)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.14.7 (tags/v3.14.7:823f032, Aug  5 2026, 10:51:32) [MSC v.1944 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-11-10.0.26200-SP0
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Is XPU available: False
Is XNNPACK available: False

CPU:
Name: 11th Gen Intel(R) Core(TM) i9-11900T @ 1.50GHz
Manufacturer: GenuineIntel
Architecture: 9

Versions of relevant libraries:
[pip3] numpy==2.5.2
[pip3] torch==2.14.0+cpu
[pip3] torchvision==0.29.0+cpu
[conda] Could not collect
```

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

### PR DESCRIPTION
**Pull request by SepehrRezaee**
> **Disclosure (PyTorch AI policy):** This PR and its description were drafted with an AI coding assistant; the diff and this text have been reviewed by the human submitter, and the PR is opened in **draft** for maintainer review.

## Description

`torch.nn.functional.pdist` accepts a zero-row input and returns an empty `(0,)` tensor, but calling `.backward()` on that output killed the process (integer division by zero: exit code `0xC0000094` on Windows, SIGSEGV on Linux) instead of raising a Python exception. Inputs with one or two rows were fine, and CUDA already returned a `(0, d)` gradient.

Root cause: `_pdist_backward` dispatched every input to the device kernels, and the CPU kernel computes its parallel grain size as `GRAIN_SIZE / (8 * n * n)` with `n = self.size(0)`, dividing by zero for `n == 0`. The kernel already documents "Assumes self is nonempty"; the forward (`_pdist_forward`) already special-cases `self.size(0) <= 1`; and `_cdist_backward` already guards empty tensors at the same dispatch level ("Gracefully handle empty Tensors").

Fix: add the same guard to `_pdist_backward` - return the (empty) result before dispatching to the kernels. This fixes all devices at once, leaves CUDA behavior unchanged, and covers the `torch.compile` path too since it dispatches to the same op.

## Test Plan

Added a zero-row regression case to `test_pdist_empty_row` (it previously only covered a 1-row input): forward shape `(0,)`, `.sum().backward()` must not crash, and the grad shape must be `(0, 3)`. Could not run it locally before submission (no build environment available); CI will run:

```
python test/test_nn.py -v -k pdist_empty_row
```

### CODE REVIEW

**Review by SepehrRezaee — COMMENTED** [review]
Bravo

**Comment by janeyx99** [comment]
Thank you for sending a PR!
To avoid you spending time preparing a PR for a feature we're not sure we want to add, our usual process is to discuss on the issue before doing the PR.
In particular, make sure that the issue is marked "actionable" before sending a PR!
See https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for more details.
